### DISCUSSION TITLE
dev containers duplicate setup process when other tools already do this

### DISCUSSION THREAD

**Discussion by krainboltgreene** [community/community#6665] [Codespaces]
Almost every devcontainer is devoted to at least one language and has instructions for setting up that language on a particular machine.

Every time the underlying languages change how they are built we now have to (hopefully) replicate that across multiple devcontainers, for example lets look at nodejs:

  - nodejs: https://github.com/microsoft/vscode-dev-containers/blob/main/containers/javascript-node/.devcontainer/base.Dockerfile
  - ruby+rails https://github.com/microsoft/vscode-dev-containers/blob/main/containers/ruby-rails/.devcontainer/Dockerfile#L13-L14
  - elixir+phoenix: https://github.com/microsoft/vscode-dev-containers/blob/main/containers/elixir-phoenix-postgres/.devcontainer/Dockerfile#L38-L42
  - python+django: https://github.com/microsoft/vscode-dev-containers/blob/main/repository-containers/github.com/django/django/.devcontainer/Dockerfile#L7-L8
  - dotnet: https://github.com/microsoft/vscode-dev-containers/blob/main/containers/dapr-dotnet/.devcontainer/Dockerfile#L5-L7
  
However, as you can see above anytime a container isn't directly based on the `javascript-node` container it has a wildly different install process.

This isn't even getting to the fact that this all becomes a waste if the user wants to run, say, fedora or redhat as their base os.

So given that this becomes an unmaintainable scenario over time (and in my opinion already has become hard) codespaces should instead look to a shared resource that already knows how to install many many platforms and has a history of community driven support: https://asdf-vm.com/

  1. An install is easy to read and easy to change: `asdf install nodejs lts`
  2. It handles the various issues with different platforms
  3. It's written by someone else

tl;dr: devcontainers doesn't need to be a duplicate, often lacking, effort to install all types of programming tools, instead it should be about mapping those tools to vscode.

Update: As if to confirm my convictions today I found out that `erlang`'s official docker image is based on buster/jessie, and doesn't work with focal. In fact the only clear and concise directions on how to upgrade were in the *asdf-erlang* documentation.

**Comment by Chuxel** [comment]
Many of the items you list are community contributed and we want to respect each ecosystem's preferences in how their images are built, the packaging systems they prefer. To your point, you do have to have knowledge of the base image to some extent since since things vary even between Debian versions. For non-community definitions, we pre-build and monitor them for both security and create performance, so these will have some additional contents that cannot be used in other scenarios. That said, we definitely know we need to provide a better way to allow external contributors to create their own dev container definitions that are shared than we have today - so big +1 here.

Independent of this, we also plan to let people contribute their own "features" that can be added to one or more "base" definitions.  @2percentsilk demoed the idea of a feature during the keynote today. These include install scripts along with configuration to add which gets the the mapping you mention. The start of this already exists, but the external contribution story is still forming. This will remove a number of the definitions that exist in this repository and things like adding node.js become a reference instead, which improves consistency. Furthermore, this allows individual feature authors can decide how to implement as they see fit along with the test matrix they are willing to take on (e.g. testing Fedora, etc). In the case of the base images we are maintain, we focus on Debian and Ubuntu primarily. You can see the start of what we should be able to do by using the **Add Development Container Configuration Files** command from VS Code. There is [also a CLI](https://code.visualstudio.com/docs/remote/devcontainer-cli) that allows for image building.

There is already a "Homebrew" feature, and certainly other packaging systems could be added as well. Insofar as a feature author wants to use one, they can opt to install it (e.g. making SDKMAN available for the Java ecosystem).  Rather than backing a single packaging system in a sea of many, we can give users the option to install their manager of choice to move forward. We've had requests to add everything from asdf to brew to Nix which all goes to show that there's a wide variety of options. We want to unlock each community's ability to rally around the things that make the most sense to them.

**Reply by krainboltgreene** [reply]
> @2percentsilk demoed the idea of a feature during the keynote today. These include install scripts along with configuration to add which gets the the mapping you mention.

The more I think about it the more I believe this to be the correct path forward. You already have a plethora of shell scripts, they're already conveniently marked with the platform they're designed for `-debian.sh`.

**Reply by krainboltgreene** [reply]
![image](https://user-images.githubusercontent.com/334809/139559492-4fdb19a7-78a0-40aa-9cdc-95308d06b0f9.png)

I've already started doing this.

**Comment by krainboltgreene** [comment]
> This will remove a number of the definitions that exist in this repository and things like adding node.js become a reference instead, which improves consistency.

Great to hear. I want to emphasize that I don't see this as a Github/Codespaces/VSCode failing but a docker failing. If only there was a way to `MERGE {image}` in a dockerfile. The idea that you can create cohesive developer environments from a "tree"-style of FROMs seems to have fallen flat on it's face in reality.

> Independent of this, we also plan to let people contribute their own "features" that can be added to one or more "base" definitions. @2percentsilk demoed the idea of a feature during the keynote today. These include install scripts along with configuration to add which gets the the mapping you mention. The start of this already exists, but the external contribution story is still forming.

Very exciting to hear about. I am personally not having a lot of fun manually updating my copies of `common-debian.sh` or `docker-debian.sh`.

> Many of the items you list are community contributed and we want to respect each ecosystem's preferences in how their images are built, the packaging systems they prefer. [...] We want to unlock each community's ability to rally around the things that make the most sense to them.

I would say you have done that excellently so far, which is also why we have 5 different install techniques for nodejs represented in the repository. Are people suggesting asdf, nix, homebrew simply because they want the end result: An easy way to install developer software, or because they care deeply about *how* they install developer software. Sometimes in feature development, as I'm sure you've experienced, having *an* answer is as meaningful as having the answer you want.
